server/providers/uploads/local/index.ts

```diff
@@ -0,0 +1,72 @@
+import { Provider } from '../../../uploads/types'
+import process from 'process'
+import path from 'path'
+import fs from 'fs'
+import crypto from 'crypto'
+import express from 'express'
+import config from '../../../../config/server'
+
+interface LocalProviderOptions {
+  uploadDirectory?: string;
+  endpoint: string;
+}
+
+interface Upload {
+  filePath: string;
+  name: string;
+}
+
+export default class LocalProvider implements Provider {
+  private uploadDirectory: string
+  private endpoint: string
+  private app: express.Application
+
+  private uploadMap: Map<string, Upload>
+
+  constructor (options: LocalProviderOptions, app: express.Application) {
+    if (options.uploadDirectory === undefined) {
+      options.uploadDirectory = path.join(process.cwd(), 'uploads')
+    }
+
+    if (!fs.existsSync(options.uploadDirectory)) {
+      fs.mkdirSync(options.uploadDirectory)
+    }
+
+    this.uploadDirectory = options.uploadDirectory
+    this.endpoint = options.endpoint
+    this.app = app
+
+    this.uploadMap = new Map()
+    this.app.get(this.endpoint, (req, res) => {
+      const key = req.query.key.toString()
+
+      if (this.uploadMap.has(key)) {
```

Comment:
Do we need to care if the file with a given hash was uploaded with the same name as it’s currently being requested as? I think a check for the file by hash existing on disk is good enough, and we can serve it under whatever name it was requested with.

Reply:
That seems like unnecessary flexibility. Ideally, the key is an opaque string that maps directly to a file/name combo.